src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -960,12 +960,6 @@ private static ValueTypeShapeCharacteristics ComputeHomogeneousAggregateCharacte
 
             MetadataType metadataType = (MetadataType)type;
 
-            // No HAs with explicit layout. There may be cases where explicit layout may be still
-            // eligible for HA, but it is hard to tell the real intent. Make it simple and just
-            // unconditionally disable HAs for explicit layout.
-            if (metadataType.IsExplicitLayout)
-                return NotHA;
```

Comment:
Will a structure like below be considered an HFA now?
```C#
[StructLayout(LayoutKind.Explicit)]
struct Foo
{
    [FieldOffset(0)] double f1;
    [FieldOffset(5)] double f2;
}
```

Reply:
Thanks Anton, you have raised a very interesting point. In the particular case you call out it won't because I think the total size will be 13 and according to line 1032 it will fail checking that the size is divisible by the floating point size, however in the loop over fields the individual fields aren't checked for the presence of proper alignment. Interestingly enough, in the original native runtime version I used for the update,

https://github.com/dotnet/runtime/blob/1928cd2b65c04ebe6fe528d4ebb581e46f1fed47/src/coreclr/vm/class.cpp#L1595

the check is present for R4 and R8 types but not for the general valuetype case so that, if I'm not mistaken, replacing the type of your <code>f2</code> with a <code>struct</code> containing a single <code>double</code> field would be erroneously considered HFA by the runtime. As a first step, I'll update this PR by adding the modulus check to the loop, we can then discuss whether the native runtime also needs fixing or if I'm just misreading the code.